lib/Stan_math_code_gen.ml

```diff
@@ -116,61 +136,115 @@ let pp_arg ppf ((adtype, name, st), idx) =
+let pp_returntype ppf arg_types rt =
+  let scalar =
+    strf "typename boost::math::tools::promote_args<@[<-35>%a@]>::type"
+      (list ~sep:comma string)
+      (maybe_templated_arg_types arg_types)
+  in
+  pf ppf "%a@ " (option ~none:(const string "void") (pp_unsizedtype scalar)) rt
 
-let pp_decl ppf (vident, st) =
-  pf ppf "%a %s;" pp_prim_stantype (Ast.remove_size st) vident
+let pp_location = fmt "current_statement__ = %S;@;"
 
-let with_no_loc stmt = {stmt; sloc= ""}
+(** [pp_located_error ppf (body_block, err_msg)] surrounds [body_block]
+    with a C++ try-catch that will rethrow the error with the proper source location
+    from the [body_block] (required to be a [stmt_loc Block] variant).*)
+let pp_located_error ppf (pp_body_block, body, err_msg) =
+  pf ppf "@ try %a" pp_body_block body ;
+  string ppf " catch (const std::exception& e) " ;
+  pp_block ppf (pp_located_msg, err_msg)
+
+let rec pp_for_each_in_array ppf (cctype, pp_body, ident) =
+  match cctype with
+  | Ast.SArray (t, dim) ->
+      let loopvar, reset_sym = Mir.gensym_enter () in
+      let new_ident = strf "%s[%s]" ident loopvar in
+      let new_pp_body ppf ident =
+        pp_for_loop ppf (loopvar, zero, dim, pp_body, ident)
+      in
+      pp_for_each_in_array ppf (t, new_pp_body, new_ident) ;
+      reset_sym ()
+  | _ -> pp_body ppf ident
+
+let%expect_test "single pp_for_each_in_array" =
+  let ppbody ppf id = pf ppf "%s++" id in
+  strf "%a" pp_for_each_in_array (Ast.SReal, ppbody, "x") |> print_endline ;
+  [%expect {| x++ |}]
+
+let%expect_test "for each in array" =
+  let ppbody ppf id = pf ppf "check_whatever(%s);" id in
+  strf "%a" pp_for_each_in_array
+    (Ast.SArray (Ast.SArray (Ast.SReal, Var "z"), Var "y"), ppbody, "alpha")
+  |> print_endline ;
+  [%expect
+    {|
+    for (size_t sym2 = 0; sym2 < z; sym2++)
+      for (size_t sym1 = 0; sym1 < y; sym1++) check_whatever(alpha[sym1][sym2]); |}]
+
+let trans_math_fn fname =
+  match fname with "print" -> ("stan_print", [Var "pstream__"]) | x -> (x, [])
 
 let rec pp_statement ppf {stmt; sloc} =
   ( match stmt with
-  | Block _ | SList _ -> ()
-  | _ -> pf ppf "current_statement_loc__ = \"%s\";@;" sloc ) ;
+  | Block _ | SList _ | FunDef _ | Break | Continue | Skip -> ()
+  | _ -> pp_location ppf sloc ) ;
   let pp_stmt_list = list ~sep:cut pp_statement in
   match stmt with
   | Assignment (assignee, rhs) ->
       (* XXX completely wrong *)
       pf ppf "%a = %a;" pp_expr assignee pp_expr rhs
   | NRFunApp (fname, args) ->
-      pf ppf "%s(%a);" fname (list ~sep:comma pp_expr) args
+      let fname, extra_args = trans_math_fn fname in
+      pf ppf "%s(@[<hov>%a@]);" fname (list ~sep:comma pp_expr)
+        (extra_args @ args)
   | Break -> string ppf "break;"
   | Continue -> string ppf "continue;"
   | Return e -> pf ppf "return %a;" (option pp_expr) e
   | Skip -> ()
-  | MarkLocation _ -> () (* XXX *)
-  | Check _ -> () (* XXX *)
+  | Check {ccfunname; ccvid; ccargs; cctype} ->
```

Comment:
Does this also do the right thing if `cctype` is an Eigen-type?

Reply:
Yep.